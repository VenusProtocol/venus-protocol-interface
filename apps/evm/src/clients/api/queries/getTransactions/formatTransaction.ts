import BigNumber from 'bignumber.js';

import { Token, TransactionCategory, TransactionEvent, VToken } from 'types';
import { findTokenByAddress } from 'utilities';

import { TransactionResponse } from './types';

const formatTransaction = ({
  data: { amountMantissa, category, event, tokenAddress, timestamp, from, ...rest },
  vTokens,
  tokens,
  defaultToken,
}: {
  data: TransactionResponse;
  vTokens: VToken[];
  tokens: Token[];
  defaultToken: Token;
}) => {
  // check if the tokenAddress is from a VToken
  const vToken = findTokenByAddress({
    address: tokenAddress || '',
    tokens: vTokens,
  });

  // if it is, use the VToken decimals and the image from the underlying token
  const transactionToken: Token | undefined = vToken
    ? { ...vToken.underlyingToken, decimals: vToken.decimals }
    : // else get the token from tokenAddress
      findTokenByAddress({ address: tokenAddress || '', tokens });

  // if neither is found, use XVS
  const token = transactionToken || defaultToken;

  return {
    ...rest,
    amountMantissa: new BigNumber(amountMantissa),
    category: category as TransactionCategory,
    event: event as TransactionEvent,
    token,
    from,
    timestamp: new Date(timestamp * 1000), // Convert timestamp to milliseconds
  };
};

export default formatTransaction;
