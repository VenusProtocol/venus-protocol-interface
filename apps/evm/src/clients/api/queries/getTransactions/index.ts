import { VError } from 'packages/errors';
import { Token, Transaction, TransactionEvent, VToken } from 'types';
import { restService } from 'utilities';

import formatTransaction from './formatTransaction';
import { TransactionResponse } from './types';

export interface GetTransactionsInput {
  vTokens: VToken[];
  tokens: Token[];
  defaultToken: Token;
  page?: number;
  event?: TransactionEvent;
  from?: string;
  sort?: '+';
  order?:
    | 'id'
    | 'event'
    | 'transactionHash'
    | 'blockNumber'
    | 'from'
    | 'to'
    | 'amountMantissa'
    | 'createdAt';
}

export interface GetTransactionsResponse {
  limit: number;
  page: number;
  total: number;
  result: TransactionResponse[];
}

export interface GetTransactionsOutput {
  limit: number;
  page: number;
  total: number;
  transactions: Transaction[];
}

const getTransactions = async ({
  vTokens,
  tokens,
  defaultToken,
  page = 0,
  event,
  from,
  order,
  sort,
}: GetTransactionsInput): Promise<GetTransactionsOutput> => {
  const orderWithSort = sort && order ? sort + order : order;
  const response = await restService<GetTransactionsResponse>({
    endpoint: '/activity/transactions',
    method: 'GET',
    params: {
      page,
      event,
      from,
      order: orderWithSort,
    },
  });
  const payload = response.data;
  // @todo Add specific api error handling
  if (payload && 'error' in payload) {
    throw new VError({
      type: 'unexpected',
      code: 'somethingWentWrong',
      data: { message: payload.error },
    });
  }

  if (!payload) {
    throw new VError({ type: 'unexpected', code: 'somethingWentWrongRetrievingTransactions' });
  }
  const { limit, page: payloadPage, total } = payload;
  const transactions = payload.result.map(data =>
    formatTransaction({ data, vTokens, defaultToken, tokens }),
  );

  return { limit, page: payloadPage, total, transactions };
};

export default getTransactions;
