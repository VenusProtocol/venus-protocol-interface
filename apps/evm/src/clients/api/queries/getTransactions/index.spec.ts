import Vi from 'vitest';

import fakeAddress from '__mocks__/models/address';
import tokens, { xvs } from '__mocks__/models/tokens';
import { transactionResponse } from '__mocks__/models/transactions';
import vTokens from '__mocks__/models/vTokens';

import { restService } from 'utilities';

import getTransactions from '.';

vi.mock('utilities/restService');

describe('api/queries/getTransactions', () => {
  it('returns transaction models', async () => {
    (restService as Vi.Mock).mockImplementationOnce(async () => ({
      status: 200,
      data: { result: transactionResponse, limit: 20, page: 1, total: 40 },
    }));

    const { transactions } = await getTransactions({
      page: 2,
      event: 'Withdraw',
      order: 'event',
      from: fakeAddress,
      sort: '+',
      vTokens,
      tokens,
      defaultToken: xvs,
    });

    expect(restService).toBeCalledWith({
      endpoint: '/activity/transactions',
      method: 'GET',
      params: {
        page: 2,
        event: 'Withdraw',
        order: '+event',
        from: fakeAddress,
      },
    });

    expect(transactions).toMatchSnapshot();
  });

  it('gets called with correct default arguments', async () => {
    (restService as Vi.Mock).mockImplementationOnce(async () => ({
      status: 200,
      data: { result: transactionResponse, limit: 20, page: 1, total: 40 },
    }));

    const { transactions } = await getTransactions({
      vTokens,
      tokens,
      defaultToken: xvs,
    });

    expect(transactions).toHaveLength(20);

    expect(restService).toBeCalledWith({
      endpoint: '/activity/transactions',
      method: 'GET',
      params: {
        page: 0,
        event: undefined,
        order: undefined,
        address: undefined,
        sort: undefined,
      },
    });

    expect(transactions).toMatchSnapshot();
  });
});
