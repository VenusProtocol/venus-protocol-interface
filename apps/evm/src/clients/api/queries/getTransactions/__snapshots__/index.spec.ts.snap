// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`api/queries/getTransactions > gets called with correct default arguments 1`] = `
[
  {
    "amountMantissa": "0",
    "blockNumber": 19636734,
    "category": "vtoken",
    "event": "Transfer",
    "from": "0x08e0A5575De71037aE36AbfAfb516595fE68e5e4",
    "logIndex": 1,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "0xD4aDbd5ed497c7720127Bfe9b05D725aC72ae2a9",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x6b8f0ebd99034cf5bec250a2e14b1d594e85be39a2cecd1193536e9576897800",
  },
  {
    "amountMantissa": "499999979046",
    "blockNumber": 19636720,
    "category": "vtoken",
    "event": "Transfer",
    "from": "0xD4aDbd5ed497c7720127Bfe9b05D725aC72ae2a9",
    "logIndex": 2,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "0x08e0A5575De71037aE36AbfAfb516595fE68e5e4",
    "token": {
      "address": "0x8301F2213c0eeD49a7E28Ae4c3e91722919B8B47",
      "asset": "/src/libs/tokens/img/busd.svg",
      "decimals": 18,
      "symbol": "BUSD",
    },
    "transactionHash": "0xb1739f27bf65398459df3228c0c74d955e8438e831fbde17506490368b264bf4",
  },
  {
    "amountMantissa": "5e+21",
    "blockNumber": 19632326,
    "category": "vtoken",
    "event": "Transfer",
    "from": "0x2E7222e51c0f6e98610A1543Aa3836E092CDe62c",
    "logIndex": 3,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "0x8eE02B3FeCcae787992c4790bc31b350E7d1F382",
    "token": {
      "address": "0xA11c8D9DC9b66E209Ef60F0C8D969D3CD988782c",
      "asset": "/src/libs/tokens/img/usdt.svg",
      "decimals": 8,
      "symbol": "USDT",
    },
    "transactionHash": "0x883ef64e9c1b043325834e1a9109bb5a31af4af67cac0b8c9d82b777c0c6efb7",
  },
  {
    "amountMantissa": "39000000000000000000",
    "blockNumber": 19625976,
    "category": "vtoken",
    "event": "Mint",
    "from": "0x7D011896EAF0D59a4Da7CD9573F4F15AB5123Ab2",
    "logIndex": 4,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x40bcf58c72dcf0f56cf349155178de5d5b34b1cc36a00ee3c52789538acbfa23",
  },
  {
    "amountMantissa": "3.33225e+24",
    "blockNumber": 19614816,
    "category": "vtoken",
    "event": "Redeem",
    "from": "0xa60dF6af57fb55c62511cA96eD6dA0AfE977436b",
    "logIndex": 5,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x389d646c3b3516eb905a571e327b4a368757f9c24963efe88e9c3cb6bacd89c4",
  },
  {
    "amountMantissa": "15495139729223118",
    "blockNumber": 19614805,
    "category": "vtoken",
    "event": "Transfer",
    "from": "0x2E7222e51c0f6e98610A1543Aa3836E092CDe62c",
    "logIndex": 6,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "0xa60dF6af57fb55c62511cA96eD6dA0AfE977436b",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0xc2fff9b3781c39d727773f8233e72a45af3d49fa967057a7f1c7724ef244882e",
  },
  {
    "amountMantissa": "1000000000000000000",
    "blockNumber": 19614713,
    "category": "vtoken",
    "event": "Redeem",
    "from": "0xa60dF6af57fb55c62511cA96eD6dA0AfE977436b",
    "logIndex": 7,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0xebc9e9d632f30fe45b7181801eb7c03e60b275c207c7d8ebb5e317a1becf9f5b",
  },
  {
    "amountMantissa": "8e+21",
    "blockNumber": 19614700,
    "category": "vtoken",
    "event": "Mint",
    "from": "0xa60dF6af57fb55c62511cA96eD6dA0AfE977436b",
    "logIndex": 8,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x862bdfa3b617562c5729b258728d2e504856c6f267bbba80fc432c3a8afe29fd",
  },
  {
    "amountMantissa": "39365649572954",
    "blockNumber": 19612497,
    "category": "vtoken",
    "event": "Mint",
    "from": "0x8eE02B3FeCcae787992c4790bc31b350E7d1F382",
    "logIndex": 9,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x85dca6ce86ae0e68b7a6486c72109322781fc4fdfd87bfccd59be258b7e7d8a1",
  },
  {
    "amountMantissa": "2036811601414153",
    "blockNumber": 19606971,
    "category": "vtoken",
    "event": "Borrow",
    "from": "0xB4A5A75A3cA76CF0939EF4A1fFb5DF3432F80e96",
    "logIndex": 10,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x496c5be16dcd2c1127f45879e0e571bb63b31cf197cbf069cf81cadf23686cfc",
  },
  {
    "amountMantissa": "2e+21",
    "blockNumber": 19606940,
    "category": "vtoken",
    "event": "Mint",
    "from": "0xB4A5A75A3cA76CF0939EF4A1fFb5DF3432F80e96",
    "logIndex": 11,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x4eeeab99581780110930b584faa5a2290c6dd989da23d67142da94aef6e2567c",
  },
  {
    "amountMantissa": "1255762298411977",
    "blockNumber": 19605600,
    "category": "vtoken",
    "event": "Transfer",
    "from": "0x08e0A5575De71037aE36AbfAfb516595fE68e5e4",
    "logIndex": 12,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "0x7b702744A7F9042670FD3d294eD81C82bA6A0350",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x000e045d8120d6de9a36e0d24bfcff2a52f8dbb2612e507fbe1d89d772a4f335",
  },
  {
    "amountMantissa": "63969210214200",
    "blockNumber": 19605119,
    "category": "vtoken",
    "event": "Mint",
    "from": "0x1A310A642C1B9Ace2e61ACbA9408DcFD27100E46",
    "logIndex": 13,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x65f3e3923876aca08d4a4b6ad2e0a985f049732ede665faacf0490b04d35144d",
  },
  {
    "amountMantissa": "68256811",
    "blockNumber": 19589102,
    "category": "vtoken",
    "event": "Transfer",
    "from": "0x2D7A3Ae4FC2aaF3852244032F6aCFd2C1b953c5B",
    "logIndex": 14,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "0xb6e9322C49FD75a367Fcb17B0Fcd62C5070EbCBe",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x406ed11513e989a8532130c88a776416d91166aede02d152500af64700f63a76",
  },
  {
    "amountMantissa": "148302657892509813122",
    "blockNumber": 19589089,
    "category": "vtoken",
    "event": "Transfer",
    "from": "0x1bD8Aa256b500f260855DE5A307b735aeF25f96a",
    "logIndex": 15,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "0x488aB2826a154da01CC4CC16A8C83d4720D3cA2C",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0xbd3324f298e8064e7cd6e4d53b322ea5778a79003abb25e010340857ac2ef5e2",
  },
  {
    "amountMantissa": "68426113815486080",
    "blockNumber": 19589043,
    "category": "vtoken",
    "event": "Mint",
    "from": "0xdD8C11CEedB110eC42b467D5861252dA9130Bf6C",
    "logIndex": 16,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x85df57add40993d57b11b203743ffb3ec7b615c684c79754c84217e645898717",
  },
  {
    "amountMantissa": "800000000000000000",
    "blockNumber": 19588967,
    "category": "vtoken",
    "event": "Transfer",
    "from": "0xb6e9322C49FD75a367Fcb17B0Fcd62C5070EbCBe",
    "logIndex": 17,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "0xdD8C11CEedB110eC42b467D5861252dA9130Bf6C",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x102add48b0ce2de31d39b45d4e1bb8f55eb45c884146607487d09857b1b0bf7b",
  },
  {
    "amountMantissa": "277449960",
    "blockNumber": 19588945,
    "category": "vtoken",
    "event": "Transfer",
    "from": "0x2E7222e51c0f6e98610A1543Aa3836E092CDe62c",
    "logIndex": 18,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "0xdD8C11CEedB110eC42b467D5861252dA9130Bf6C",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x839f73ce52cd04472dcea7e7b6670d3027ebd309e9bd7170345965d96ffed914",
  },
  {
    "amountMantissa": "159200000000000000000",
    "blockNumber": 19588667,
    "category": "vtoken",
    "event": "Redeem",
    "from": "0x6DffAF09502CD4a9DfcB09436593D13aeC921ffD",
    "logIndex": 19,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x716d01587f26bbe313dab9858756f32c79efd5b0d33609e8ad7a294dfa7a6d31",
  },
  {
    "amountMantissa": "3936585693",
    "blockNumber": 19588665,
    "category": "vtoken",
    "event": "Transfer",
    "from": "0x8485F50a695262C84f06679Ea9Ea697Aae7227dE",
    "logIndex": 20,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "0x08e0A5575De71037aE36AbfAfb516595fE68e5e4",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x51cea410036b49398541f309c855acd8ec51f3ea802a61a85edd0af67944767d",
  },
]
`;

exports[`api/queries/getTransactions > returns transaction models 1`] = `
[
  {
    "amountMantissa": "0",
    "blockNumber": 19636734,
    "category": "vtoken",
    "event": "Transfer",
    "from": "0x08e0A5575De71037aE36AbfAfb516595fE68e5e4",
    "logIndex": 1,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "0xD4aDbd5ed497c7720127Bfe9b05D725aC72ae2a9",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x6b8f0ebd99034cf5bec250a2e14b1d594e85be39a2cecd1193536e9576897800",
  },
  {
    "amountMantissa": "499999979046",
    "blockNumber": 19636720,
    "category": "vtoken",
    "event": "Transfer",
    "from": "0xD4aDbd5ed497c7720127Bfe9b05D725aC72ae2a9",
    "logIndex": 2,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "0x08e0A5575De71037aE36AbfAfb516595fE68e5e4",
    "token": {
      "address": "0x8301F2213c0eeD49a7E28Ae4c3e91722919B8B47",
      "asset": "/src/libs/tokens/img/busd.svg",
      "decimals": 18,
      "symbol": "BUSD",
    },
    "transactionHash": "0xb1739f27bf65398459df3228c0c74d955e8438e831fbde17506490368b264bf4",
  },
  {
    "amountMantissa": "5e+21",
    "blockNumber": 19632326,
    "category": "vtoken",
    "event": "Transfer",
    "from": "0x2E7222e51c0f6e98610A1543Aa3836E092CDe62c",
    "logIndex": 3,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "0x8eE02B3FeCcae787992c4790bc31b350E7d1F382",
    "token": {
      "address": "0xA11c8D9DC9b66E209Ef60F0C8D969D3CD988782c",
      "asset": "/src/libs/tokens/img/usdt.svg",
      "decimals": 8,
      "symbol": "USDT",
    },
    "transactionHash": "0x883ef64e9c1b043325834e1a9109bb5a31af4af67cac0b8c9d82b777c0c6efb7",
  },
  {
    "amountMantissa": "39000000000000000000",
    "blockNumber": 19625976,
    "category": "vtoken",
    "event": "Mint",
    "from": "0x7D011896EAF0D59a4Da7CD9573F4F15AB5123Ab2",
    "logIndex": 4,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x40bcf58c72dcf0f56cf349155178de5d5b34b1cc36a00ee3c52789538acbfa23",
  },
  {
    "amountMantissa": "3.33225e+24",
    "blockNumber": 19614816,
    "category": "vtoken",
    "event": "Redeem",
    "from": "0xa60dF6af57fb55c62511cA96eD6dA0AfE977436b",
    "logIndex": 5,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x389d646c3b3516eb905a571e327b4a368757f9c24963efe88e9c3cb6bacd89c4",
  },
  {
    "amountMantissa": "15495139729223118",
    "blockNumber": 19614805,
    "category": "vtoken",
    "event": "Transfer",
    "from": "0x2E7222e51c0f6e98610A1543Aa3836E092CDe62c",
    "logIndex": 6,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "0xa60dF6af57fb55c62511cA96eD6dA0AfE977436b",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0xc2fff9b3781c39d727773f8233e72a45af3d49fa967057a7f1c7724ef244882e",
  },
  {
    "amountMantissa": "1000000000000000000",
    "blockNumber": 19614713,
    "category": "vtoken",
    "event": "Redeem",
    "from": "0xa60dF6af57fb55c62511cA96eD6dA0AfE977436b",
    "logIndex": 7,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0xebc9e9d632f30fe45b7181801eb7c03e60b275c207c7d8ebb5e317a1becf9f5b",
  },
  {
    "amountMantissa": "8e+21",
    "blockNumber": 19614700,
    "category": "vtoken",
    "event": "Mint",
    "from": "0xa60dF6af57fb55c62511cA96eD6dA0AfE977436b",
    "logIndex": 8,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x862bdfa3b617562c5729b258728d2e504856c6f267bbba80fc432c3a8afe29fd",
  },
  {
    "amountMantissa": "39365649572954",
    "blockNumber": 19612497,
    "category": "vtoken",
    "event": "Mint",
    "from": "0x8eE02B3FeCcae787992c4790bc31b350E7d1F382",
    "logIndex": 9,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x85dca6ce86ae0e68b7a6486c72109322781fc4fdfd87bfccd59be258b7e7d8a1",
  },
  {
    "amountMantissa": "2036811601414153",
    "blockNumber": 19606971,
    "category": "vtoken",
    "event": "Borrow",
    "from": "0xB4A5A75A3cA76CF0939EF4A1fFb5DF3432F80e96",
    "logIndex": 10,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x496c5be16dcd2c1127f45879e0e571bb63b31cf197cbf069cf81cadf23686cfc",
  },
  {
    "amountMantissa": "2e+21",
    "blockNumber": 19606940,
    "category": "vtoken",
    "event": "Mint",
    "from": "0xB4A5A75A3cA76CF0939EF4A1fFb5DF3432F80e96",
    "logIndex": 11,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x4eeeab99581780110930b584faa5a2290c6dd989da23d67142da94aef6e2567c",
  },
  {
    "amountMantissa": "1255762298411977",
    "blockNumber": 19605600,
    "category": "vtoken",
    "event": "Transfer",
    "from": "0x08e0A5575De71037aE36AbfAfb516595fE68e5e4",
    "logIndex": 12,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "0x7b702744A7F9042670FD3d294eD81C82bA6A0350",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x000e045d8120d6de9a36e0d24bfcff2a52f8dbb2612e507fbe1d89d772a4f335",
  },
  {
    "amountMantissa": "63969210214200",
    "blockNumber": 19605119,
    "category": "vtoken",
    "event": "Mint",
    "from": "0x1A310A642C1B9Ace2e61ACbA9408DcFD27100E46",
    "logIndex": 13,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x65f3e3923876aca08d4a4b6ad2e0a985f049732ede665faacf0490b04d35144d",
  },
  {
    "amountMantissa": "68256811",
    "blockNumber": 19589102,
    "category": "vtoken",
    "event": "Transfer",
    "from": "0x2D7A3Ae4FC2aaF3852244032F6aCFd2C1b953c5B",
    "logIndex": 14,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "0xb6e9322C49FD75a367Fcb17B0Fcd62C5070EbCBe",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x406ed11513e989a8532130c88a776416d91166aede02d152500af64700f63a76",
  },
  {
    "amountMantissa": "148302657892509813122",
    "blockNumber": 19589089,
    "category": "vtoken",
    "event": "Transfer",
    "from": "0x1bD8Aa256b500f260855DE5A307b735aeF25f96a",
    "logIndex": 15,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "0x488aB2826a154da01CC4CC16A8C83d4720D3cA2C",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0xbd3324f298e8064e7cd6e4d53b322ea5778a79003abb25e010340857ac2ef5e2",
  },
  {
    "amountMantissa": "68426113815486080",
    "blockNumber": 19589043,
    "category": "vtoken",
    "event": "Mint",
    "from": "0xdD8C11CEedB110eC42b467D5861252dA9130Bf6C",
    "logIndex": 16,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x85df57add40993d57b11b203743ffb3ec7b615c684c79754c84217e645898717",
  },
  {
    "amountMantissa": "800000000000000000",
    "blockNumber": 19588967,
    "category": "vtoken",
    "event": "Transfer",
    "from": "0xb6e9322C49FD75a367Fcb17B0Fcd62C5070EbCBe",
    "logIndex": 17,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "0xdD8C11CEedB110eC42b467D5861252dA9130Bf6C",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x102add48b0ce2de31d39b45d4e1bb8f55eb45c884146607487d09857b1b0bf7b",
  },
  {
    "amountMantissa": "277449960",
    "blockNumber": 19588945,
    "category": "vtoken",
    "event": "Transfer",
    "from": "0x2E7222e51c0f6e98610A1543Aa3836E092CDe62c",
    "logIndex": 18,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "0xdD8C11CEedB110eC42b467D5861252dA9130Bf6C",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x839f73ce52cd04472dcea7e7b6670d3027ebd309e9bd7170345965d96ffed914",
  },
  {
    "amountMantissa": "159200000000000000000",
    "blockNumber": 19588667,
    "category": "vtoken",
    "event": "Redeem",
    "from": "0x6DffAF09502CD4a9DfcB09436593D13aeC921ffD",
    "logIndex": 19,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x716d01587f26bbe313dab9858756f32c79efd5b0d33609e8ad7a294dfa7a6d31",
  },
  {
    "amountMantissa": "3936585693",
    "blockNumber": 19588665,
    "category": "vtoken",
    "event": "Transfer",
    "from": "0x8485F50a695262C84f06679Ea9Ea697Aae7227dE",
    "logIndex": 20,
    "timestamp": +055130-07-03T03:04:29.000Z,
    "to": "0x08e0A5575De71037aE36AbfAfb516595fE68e5e4",
    "token": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "transactionHash": "0x51cea410036b49398541f309c855acd8ec51f3ea802a61a85edd0af67944767d",
  },
]
`;
